### PR TITLE
Add a checksum for UDP messages

### PR DESCRIPTION
**Pull request by kyhavlov**
Adds a checksum to detect UDP corruption as described in hashicorp/consul#2519.

### CODE REVIEW

File: memberlist.go [hunk 1]
```diff
@@ -495,8 +495,18 @@ func (m *Memberlist) SendTo(to net.Addr, msg []byte) error {
 	buf[0] = byte(userMsg)
 	buf = append(buf, msg...)
 
+	// Lookup node from address
+	toAddr := strings.Split(to.String(), ":")[0]
```
Comment (slackpad) [inline memberlist.go]: This'll crash if the split doesn't work - let's use https://golang.org/pkg/net/#SplitHostPort and check the error return from that.

File: memberlist_test.go [hunk 2]
```diff
@@ -367,6 +367,8 @@ func (h dnsHandler) ServeDNS(w dns.ResponseWriter, r *dns.Msg) {
 }
 
 func TestMemberList_ResolveAddr_TCP_First(t *testing.T) {
+	t.Skip()
```
Comment (slackpad) [inline memberlist_test.go]: This is a leftover I think.

File: net.go [hunk 3]
```diff
@@ -599,9 +613,19 @@ func (m *Memberlist) sendMsg(to net.Addr, msg []byte) error {
 	}
 	extra := m.getBroadcasts(compoundOverhead, bytesAvail)
 
+	// Lookup node from address
+	toAddr := strings.Split(to.String(), ":")[0]
```
Comment (slackpad) [inline net.go]: Same for this one as above.

File: net.go [hunk 4]
```diff
@@ -338,8 +341,19 @@ func (m *Memberlist) ingestPacket(buf []byte, from net.Addr, timestamp time.Time
 		buf = plain
 	}
 
-	// Handle the command
-	m.handleCommand(buf, from, timestamp)
+	// See if there's a checksum included to verify the contents of the message
+	msgType := messageType(buf[0])
```
Comment (slackpad) [inline net.go]: Need a size check to make sure buf isn't empty, otherwise this can panic.

Comment (slackpad) [inline net.go]: You actually need to check for at least length 5.

File: net.go [hunk 5]
```diff
@@ -338,8 +341,19 @@ func (m *Memberlist) ingestPacket(buf []byte, from net.Addr, timestamp time.Time
...
+	if msgType == hasCrcMsg {
+		crc := crc32.ChecksumIEEE(buf[5:])
+		expected := binary.BigEndian.Uint32(buf[1:5])
+		if crc != expected {
+			m.logger.Printf("[WARN] memberlist: Got invalid checksum for UDP packet: %d, %d", crc, expected)
```
Comment (slackpad) [inline net.go]: Print as hex is probably easier to interpret.

File: net.go [hunk 6]
```diff
@@ -613,11 +637,11 @@ func (m *Memberlist) sendMsg(to net.Addr, msg []byte) error {
 	compound := makeCompoundMessage(msgs)
 
 	// Send the message
-	return m.rawSendMsgUDP(to, compound.Bytes())
+	return m.rawSendMsgUDP(to, node, compound.Bytes())
 }
 
 // rawSendMsgUDP is used to send a UDP message to another host without modification
```
Comment (slackpad) [inline net.go]: Let's add a comment about it being ok for the node to be `nil`.

File: net.go [hunk 7]
```diff
@@ -631,6 +655,16 @@ func (m *Memberlist) rawSendMsgUDP(to net.Addr, msg []byte) error {
 		}
 	}
 
+	// Add a CRC to the end of the payload if the recipient understands
+	// ProtocolVersion >= 5
+	if node != nil && node.PMax >= 5 {
+		crc := crc32.ChecksumIEEE(msg)
+		sum := make([]byte, 5)
```
Comment (slackpad) [inline net.go]: `sum` is kind of a confusing name, how about `header`?

File: memberlist_test.go (hunk 2 above)
Comment (kyhavlov) [inline memberlist_test.go]: This test was already failing in master, so I had it skipped before getting around to fixing it in this PR

**Comment by slackpad** [comment]
LGTM

**Pull request merged**